library/blob.c

```diff
@@ -491,6 +695,23 @@ static int eblob_check_disk_one(struct eblob_iterate_local *loc)
 		}
 	}
 
+	if ((ctl->flags & EBLOB_ITERATE_FLAGS_VERIFY_CHECKSUM) &&
+	    !(dc->flags & BLOB_DISK_CTL_REMOVE) &&
+	    !(dc->flags & BLOB_DISK_CTL_UNCOMMITTED)) {
+		struct eblob_write_control wc;
+		memset(&wc, 0, sizeof(wc));
+		eblob_rctl_to_wc(&rc, &wc);
+		eblob_dc_to_wc(dc, &wc);
+
+		err = eblob_verify_checksum(bc->back, &dc->key, &wc);
+		if (err) {
+			eblob_dump_wc(bc->back, &dc->key, &wc, "eblob_check_disk_one: checksum verification failed", err);
+			eblob_mark_entry_removed_purge(bc->back, &dc->key, &rc);
```

Comment:
If current recovery didn't remove those keys yet, it didn't try to use these replicas for recovery. As soon as it try to use it, it will try to read them with checksum verification, after that they will be overwritten (if there is another valid replica with valid checksum) or removed (in other cases). With these changes nothing is going to be changed, it will remove only keys that it tries to recover.

What are you talking about? Eblob read call does verify checksum and compare headers. If checksum verification is failed, data is not valid. If you have broken blobs with valid data but invalid flags or missed checksums, fix them - convert to valid blobs.

Let's add recovery of headers. What other conditions do you have?